### PR TITLE
Configuration

### PR DESCRIPTION
**Pull request by przepompownia**
In order to keep all currently desired configuration rules clear I made several small steps decoupling some responsibilities such as logic of objects creation or regex pattern matching (which will be needed in class info repository). 

As you see, among many steps I did not decide to make one big class whose object could serve the whole configuration. At this moment I do not have any good idea than keeping the configuration in array at the start but I do not need such idea so much.

### CODE REVIEW

File: ftplugin/php_phpcd.vim [hunk 1]
```diff
@@ -43,12 +43,16 @@ let s:phpcd_path = expand('<sfile>:p:h:h') . '/php/main.php'
 if g:phpcd_channel_id != -1
 	call rpcstop(g:phpcd_channel_id)
 endif
-let g:phpcd_channel_id = rpcstart(g:phpcd_php_cli_executable, [s:phpcd_path, s:root, 'PHPCD'])
+
+let s:encoded_options = json_encode(g:phpcd_options)
```
Comment (lvht) [inline ftplugin/php_phpcd.vim]: Do NeoVim has the `json_encode` function ?

Comment (przepompownia) [inline ftplugin/php_phpcd.vim]: I checked: both traditional Vim and Neovim supports `json_encode`.

**Comment by przepompownia** [comment]
Standard logger implementation stays default.

File: ftplugin/php_phpcd.vim (hunk 1 above)
Comment (lvht) [inline ftplugin/php_phpcd.vim]: image

Comment (lvht) [inline ftplugin/php_phpcd.vim]:
```
NVIM 0.1.3
Build type: RelWithDebInfo
Compilation: /usr/local/Library/ENV/4.3/clang -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-u
nused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAV
E_CONFIG_H -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/build/config -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/src -I/tmp/neovim2016
0412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/tmp/neovim201
60412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include/luajit-2.0 -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/t
mp/neovim20160412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/
tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/deps-build/usr/include -I/usr/local/opt/gettext/include -I/usr/include -I/usr/include -I/tmp/neo
vim20160412-60992-1v48j6/neovim-0.1.3/build/src/nvim/auto -I/tmp/neovim20160412-60992-1v48j6/neovim-0.1.3/build/include
Compiled by lvht@192.168.1.4

Optional features included (+) or not (-): +acl   +iconv    +jemalloc
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.1.3/share/nvim"
```

**Comment by przepompownia** [comment]
Let's see to the current master and tag `v0.1.3`. It suggests me (I am not C/C++ programmer) that in 1.3 function `json_encode` is only mentioned in documentation and not implemented yet, and in the current master we can find something in `./src/nvim/eval.c`. 
```
 $ git checkout master
Previous HEAD position was f4f80a2... NVIM v0.1.3
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
 $ git log -1
commit 3098b18a2b63a841351f6d5e3697cb69db3035ef
Author: James McCoy <jamessan@jamessan.com>
Date:   Thu Apr 21 21:56:20 2016 -0400

    vim-patch.sh: Query git for name of neovim remote
    
    Rather than assume the user named their neovim/neovim remote "upstream",
    parse the information from "git remote -v".
 $ grep -r json_encode .
./src/nvim/eval/encode.c:/// Check whether given key can be used in json_encode()
./src/nvim/eval.c:  { "json_encode",        1, 1, f_json_encode },
./src/nvim/eval.c:/// json_encode() function
./src/nvim/eval.c:static void f_json_encode(typval_T *argvars, typval_T *rettv)
...
...
./runtime/syntax/php.vim:syn keyword phpFunctions  json_decode json_encode json_last_error contained
./runtime/autoload/phpcomplete.vim:let php_builtin['functions']['json']={'json_decode(':'string $json [, bool $assoc = false [, int $depth = 512 [, int $options = 0]]] | mixed','json_encode(':'mixed $value [, int $options = 0 [, int $depth = 512]] | string','json_last_error_msg(':'void | string','json_last_error(':'void | int',}
./runtime/doc/vim_diff.txt:|json_encode()| behaviour slightly changed: now |msgpack-special-dict| values 
./runtime/doc/eval.txt:         |json_encode()|.  This value is converted to "false" when used 
./runtime/doc/eval.txt:         See |json_encode()|.  This value is converted to "null" when 
./runtime/doc/eval.txt:         |json_encode()|.  This value is converted to "true" when used 
./runtime/doc/eval.txt:json_encode( {expr})             String  Convert {expr} to JSON
./runtime/doc/eval.txt:json_encode({expr})                                      *json_encode()*
./runtime/doc/eval.txt:         method, use |msgpackdump()| or |json_encode()| if you need to 
 $ git checkout v0.1.3 
Note: checking out 'v0.1.3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at f4f80a2... NVIM v0.1.3
 $ grep -r json_encode .
./runtime/syntax/php.vim:syn keyword phpFunctions  json_decode json_encode json_last_error contained
./runtime/autoload/phpcomplete.vim:let php_builtin['functions']['json']={'json_decode(':'string $json [, bool $assoc = false [, int $depth = 512 [, int $options = 0]]] | mixed','json_encode(':'mixed $value [, int $options = 0 [, int $depth = 512]] | string','json_last_error_msg(':'void | string','json_last_error(':'void | int',}
./runtime/doc/eval.txt:         method, use |msgpackdump()| or |json_encode()| if you need to 
```
I use unstable neovim version from repository which is updated almost every day (http://ppa.launchpad.net/neovim-ppa/unstable/ubuntu):
```
:version     
NVIM 0.1.4-dev                                                                                                                                                                                                                                                                    
Build type: RelWithDebInfo                                                                                                                                                                                                                                                        
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-pro
tector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/build/config -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04
.1/src -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/usr/include/luajit-2.0 -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+
git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/build/neovim-pXzt
es/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/.deps/usr/include -I/usr/include -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.1/build/src/nvim/auto -I/build/neovim-pXztes/neovim-0.1.2ubuntu1+git201604181817+2414+19~ubuntu16.04.
1/build/include                                                                                                                                                                                                                                                                   
Compiled by buildd@lgw01-27                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                  
Optional features included (+) or not (-): +acl   +iconv    +jemalloc                                                                                                                                                                                                             
For differences from Vim, see :help vim-differences                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                  
   system vimrc file: "$VIM/sysinit.vim"                                                                                                                                                                                                                                          
  fall-back for $VIM: "/usr/share/nvim"                                                                                                                                                                                                                                                                                             ```

**Comment by przepompownia** [comment]
In this case we would merge server-side changes into master and either wait with client changes for official 1.4 release or merge client side also but require the recent unstable build for latest commit. For the first option we can create special branch and put client-side changes into it.

**Comment by przepompownia** [comment]
Until `1.4` will not come out we can change `$default_options` in `main.php`.

**Comment by przepompownia** [comment]
With the last change vim should ignore vim-side configuration of server (i.e. send empty string) if it  has no `json_encode` support. It allows to merge client changes without fear of errors.

**Comment by przepompownia** [comment]
With the last change vim should ignore vim-side configuration of server (i.e. send empty string) if it  has no `json_encode` support. It allows to merge client changes without fear of errors.

**Comment by przepompownia** [comment]
I deleted this branch as it was mostly included in complete-class-names.

**Pull request closed without merging**